File: internal/godex/roadmap/validator.go
package roadmap

import (
	"fmt"
	"path/filepath"
	"strings"
)

func Validate(doc Document) []Diagnostic {
	var diagnostics []Diagnostic
	add := func(line int, message string) {
		diagnostics = append(diagnostics, Diagnostic{Path: doc.Path, Line: line, Severity: "error", Message: message})
	}
	clean := filepath.ToSlash(doc.Path)
	if clean != "" && !strings.Contains(clean, "/roadmap/") && !strings.HasPrefix(clean, "roadmap/") {
		add(0, "roadmap path must be under roadmap/")
	}
	if strings.TrimSpace(doc.Goal) == "" {
		add(0, "missing goal")
	}
	if strings.TrimSpace(doc.Architecture) == "" {
		add(0, "missing architecture")
	}
	if !containsLine(doc.Lines, "## Owned Paths") {
		add(0, "missing owned paths")
	}
	if !containsLine(doc.Lines, "## Tasks") {
		add(0, "missing tasks section")
	}
	if len(doc.Tasks) == 0 {
		add(0, "missing task checkboxes")
	}
	if len(doc.RunBlocks) == 0 || !containsLinePrefix(doc.Lines, "Run:") {
		add(0, "missing run command")
	}
	if len(doc.AcceptanceSections) == 0 {
		add(0, "missing acceptance criteria")
	}
	seen := map[string]int{}
	for _, task := range doc.Tasks {
		if first, ok := seen[task.ID]; ok {
			add(task.Line, fmt.Sprintf("duplicate task id %s first seen on line %d", task.ID, first))
			continue
		}
		seen[task.ID] = task.Line
	}
	return diagnostics
}

func containsLine(lines []string, value string) bool {
	for _, line := range lines {
		if strings.TrimSpace(line) == value {
			return true
		}
	}
	return false
}

func containsLinePrefix(lines []string, prefix string) bool {
	for _, line := range lines {
		if strings.HasPrefix(strings.TrimSpace(line), prefix) {
			return true
		}
	}
	return false
}

File: crates/roder-roadmap/src/validator.rs
use std::collections::HashSet;

use crate::parser::path_diagnostic;
use crate::{Diagnostic, DiagnosticSeverity, Document, ValidationResult};

pub fn validate_document(document: &Document) -> ValidationResult {
    let mut diagnostics = Vec::new();
    if let Some(diagnostic) = path_diagnostic(&document.path) {
        diagnostics.push(diagnostic);
    }
    require(
        &mut diagnostics,
        document,
        !document.title.trim().is_empty(),
        None,
        "missing title heading",
    );
    require(
        &mut diagnostics,
        document,
        !document.goal.trim().is_empty(),
        None,
        "missing **Goal:** field",
    );
    require(
        &mut diagnostics,
        document,
        !document.architecture.trim().is_empty(),
        None,
        "missing **Architecture:** field",
    );
    require(
        &mut diagnostics,
        document,
        !document.owned_paths.is_empty(),
        None,
        "missing owned paths",
    );
    require(
        &mut diagnostics,
        document,
        !document.tasks.is_empty(),
        None,
        "missing task checklist items",
    );
    require(
        &mut diagnostics,
        document,
        document
            .tasks
            .iter()
            .any(|task| !task.run_blocks.is_empty()),
        None,
        "missing Run block",
    );
    require(
        &mut diagnostics,
        document,
        !document.acceptance.is_empty(),
        None,
        "missing acceptance checklist",
    );

    let mut seen = HashSet::new();
    for task in &document.tasks {
        if !seen.insert(task.id.clone()) {
            diagnostics.push(Diagnostic {
                path: document.path.clone(),
                line: Some(task.line),
                severity: DiagnosticSeverity::Error,
                message: format!("duplicate task id: {}", task.id),
            });
        }
    }

    ValidationResult {
        document_id: document.id.clone(),
        diagnostics,
    }
}

fn require(
    diagnostics: &mut Vec<Diagnostic>,
    document: &Document,
    ok: bool,
    line: Option<usize>,
    message: &str,
) {
    if !ok {
        diagnostics.push(Diagnostic {
            path: document.path.clone(),
            line,
            severity: DiagnosticSeverity::Error,
            message: message.to_string(),
        });
    }
}
